librz/core/cbin.c

```diff
@@ -33,34 +33,54 @@
 
 static RZ_NULLABLE RZ_BORROW const RzList *core_bin_strings(RzCore *r, RzBinFile *file);
 
-static void pair(const char *key, const char *val) {
+static void pair(const char *key, const char *val, RzList *row_list, RzTable *t) {
 	if (!val || !*val) {
 		return;
 	}
-	rz_cons_printf("%-" PAIR_WIDTH "s%s\n", key, val);
+	RzTableColumnType *typeString = rz_table_type("string");
+	rz_table_add_column(t, typeString, key, 0);
+	rz_list_append(row_list, strdup(val));
+	val = NULL;
+	key = NULL;
+}
+
+static char *get_coloured(const char *flag, bool use_color) {
+	if (use_color == true) {
+		if (strncmp(flag, "true", 5) == 0) {
+			return rz_str_newf(Color_GREEN "true" Color_RESET);
+		} else {
+			return rz_str_newf(Color_RED "false" Color_RESET);
+		}
+	} else {
+		return strncmp(flag, "true", 5) == 0 ? strdup("true") : strdup("false");
+	}
 }
 
-static void pair_bool(PJ *pj, const char *key, bool val) {
+static void pair_bool(bool use_color, RzTable *t, PJ *pj, const char *key, bool val, RzList *row_list) {
 	if (pj) {
 		pj_kb(pj, key, val);
 	} else {
-		pair(key, rz_str_bool(val));
+		RzTableColumnType *typeString = rz_table_type("bool");
+		const char *b = val || typeString ? get_coloured(rz_str_bool(val), use_color) : "";
+		pair(key, b, row_list, t);
 	}
 }
 
-static void pair_int(PJ *pj, const char *key, int val) {
+static void pair_int(RzTable *t, PJ *pj, const char *key, int val, RzList *row_list) {
 	if (pj) {
 		pj_ki(pj, key, val);
 	} else {
-		pair(key, sdb_fmt("%d", val));
+		const char *fmt = rz_str_newf("%d", val);
+		pair(key, fmt, row_list, t);
 	}
 }
 
-static void pair_ut64(PJ *pj, const char *key, ut64 val) {
+static void pair_ut64(RzTable *t, PJ *pj, const char *key, ut64 val, RzList *row_list) {
 	if (pj) {
 		pj_kn(pj, key, val);
 	} else {
-		pair(key, sdb_fmt("%" PFMT64d, val));
+		const char *fmt = rz_str_newf("%" PFMT64d, val);
```

Comment:
`rz_str_newf` allocates memory, but i do not see you deallocating it. be sure to not produce memleaks.

you can check if you have any, via `valgrind --leak-check=full <binary path> <arguments>` (beware that you may see others from other parts of the code, but do not worry about them).

Reply:
> `rz_str_newf` allocates memory, but i do not see you deallocating it. be sure to not produce memleaks.
> 
> you can check if you have any, via `valgrind --leak-check=full <binary path> <arguments>` (beware that you may see others from other parts of the code, but do not worry about them).

```
==36510== 
--36510-- warning: DiCfSI 0x94a09d1 .. 0x99c3677 is huge; length = 5385383 (librz_asm.so.0.3.0-git)
--36510-- warning: DiCfSI 0xd34ad61 .. 0xd86da07 is huge; length = 5385383 (librz_analysis.so.0.3.0-git)
==36510==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
==36510== 
==36510== HEAP SUMMARY:
==36510==     in use at exit: 0 bytes in 0 blocks
==36510==   total heap usage: 0 allocs, 0 frees, 0 bytes allocated
==36510== 
==36510== All heap blocks were freed -- no leaks are possible
==36510== 
==36510== For lists of detected and suppressed errors, rerun with: -s
==36510== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
Valgrind gives me this ^ 

This is after I made the strdup changes. Should I change str_newf everywhere in the code or just here? @wargio

Reply:
you can use str_newf when you need to return a formatted allocated string, but if your string is a constant and you need to allocate it, then you need to use strdup

example: `char *type = strdup("bool")` and `int x = 100;  char *strnum = str_newf("%d", x);`